[UI] Refactor CurrencyEntryBox

ichthus1604 opened the pull request:
Okay, this PR is quite big.

## Description

Both `CurrencyEntryBox` and `DualCurrencyEntryBox` have been **completely rewritten**. 

The following changes have been made:

 - a `CurrencyFormat` model has been introduced, which *Represents a specific Currency's Parsing and Formatting rules*. 
    - This also defines aspects such as CurrencyCode, MaxFractionalDigits, MaxIntegralDigits, etc.
    - there are 3 CurrencyFormats right now: `CurrencyFormat.Btc`, CurrencyFormat.Usd` and `CurrencyFormat.SatvByte`
    - More could be introduced in the future to support other currencies.
    
 - `CurrencyEntryBox` now delegates all of the above aspects to the model, and deals only with user input, caret positioning, etc. It only deals with text and `decimal` Value property. it is up to the ViewModel to convert this numeric data into whatever it needs.

 - Both `CurrencyEntryBox` and `DualCurrencyEntryBox` are completely agnostic to aspects such as conversion rates, currencies, etc.

 - The clipboard handling logic has been moved to `CurrencyEntryBoxClipboardBehavior`. Once again this only deals with input and delegates logic to the `CurrencyEntryBox.CurrencyFormat` for parsing, etc.
   - The max amount logic depends on `CurrencyEntryBox.MaxValue` which gets updated (for instance in the Send View) accoding to the available balance and exchange rate.  
 - `ClipboardObserver` class has been removed.

 - `DualCurrencyEntryBox` no longer has a `readonly` state. For cases where `IsFixedAmount=true` in the send dialog, the entry box gets replaced by a simple `AmountControl` which shows the amount as readonly text.

- 2 new ViewModels have been introduced in the send screen:
  - `CurrencyViewModel` deals with a single currency input and its corresponding `MaxValue` (both in decimal format)
  - `CurrencyConversionViewModel` deals with `BTC <-> USD` conversion and also with left/right switching.

## Results

- The above makes the design of `DualCurrencyEntryBox` extremely simple. It is just a container for the 2 currency inputs and the toggle button. It has only 63 lines of code now (down from 471 on master). The switching from left to right happens in the *data*, not the *UI*, since the UI elements are currency-agnostic, this can be done.

 - The general design is **much** cleaner. Spaghetti has been removed everywhere, and now each layer (UI, Models, ViewModel) deals only with its corresponding responsibility. 
 
 - Groundwork has been laid to enable supporting multiple currencies (other than USD) and also localization-sensitive aspects have been moved to `CurrencyLocalization` class.

 - ViewModels no longer need to deal with money amounts as strings but rather as decimal amounts, regardless of currency. This also helped clean up some logic in the VMs such as format validation, etc.

 - `SendViewModel` no longer deals with clipboard contents

 - `DualCurrencyEntryBox` now has only 6 lines of logic, dealing with focus.

- The following manual tests have been made:
  - text input, including selection
  - typing "." with an empty box results in "0."
  - caret positioning affected by formatting rules
  - not being able to enter more than the max number of integral digits and fractional digits (defined by `CurrencyFormat`)
  - clipboard logic
     - paste valid values for both currencies
     - paste invalid values for both currencies
     - paste non numerical text
  - clipboard auto suggestion logic:
     - for valid numerical values
     - for invalid numerical values
     - for valid amounts within the range defined by `MaxValue` property (bound to the available balance in the Send view)
     - for invalid amounts as per the above
  - left/right switching and its effects on keyboard focus
  - readonly amount in the send dialog
  - BTC only input in the send dialog when exchange rate is not available
  - standalone uses of `CurrencyEntryBox`:
     - `CustomFeeRateDialog`
     - `PlebStopThreshold`
     - `DustThreshold`

## Problems

The only outstanding problem I couldn't solve is that for some reason the data validations are not fired when editing currency values, therefore I had to resort to a temporary hack using a hidden TextBox. @wieslawsoltes I'm going to need your help to figure this one out.

Other than that, there are stylistic things in the code, such as `CurrencyConversionViewModel` relying too much on the `_isUpdating` flag. Maybe @SuperJMN can come up with a cleaner way to handle this using Rx.

## Next steps

 - Further testing
 - Headless Avalonia based testing
 - Fixing DataValidation problem and removing the hack

turbolay commented:
> All the input bugs and usability problems that appeared after Avalonia V11 migration are gone.

Quick test: true, really good.

I have found 2 problems with pasting:
1) The cursor is not set at the end of the number when the number pasted has more than 4 digits

https://github.com/zkSNACKs/WalletWasabi/assets/16029533/13c93bd0-c7cc-487d-86e5-b1205ab015cd

(note that this doesn't happen if the string pasted already includes a space)

2) When a string has more than 8 digits, pasting doesn't work. The legacy behavior was to truncate, you can also consider this: #11802

ichthus1604 commented:
@turbolay 

> The cursor is not set at the end of the number when the number pasted has more than 4 digits

Fixed in lastest commit ;)

> When a string has more than 8 digits, pasting doesn't work. The legacy behavior was to truncate, you can also consider this: https://github.com/zkSNACKs/WalletWasabi/issues/11802

@soosr what do you think? Should we truncate or round?

soosr commented:
> > When a string has more than 8 digits, pasting doesn't work. The legacy behavior was to truncate, you can also consider this: #11802
> 
> @soosr what do you think? Should we truncate or round?

As part of this PR, I would insist on the legacy behavior. https://github.com/zkSNACKs/WalletWasabi/issues/11802 is still in Concept&Research.

Regarding the Localization, Discussed it with Adam and we should not do it. If `CurrencyInput` doesn't support the different group separators, then we need to implement it. An idea is to take the last not numeric character as the decimal separator and all the others are group separators.

soosr commented:
> Regarding the Localization, Discussed it with Adam and we should not do it. If `CurrencyInput` doesn't support the different group separators, then we need to implement it. An idea is to take the last not numeric character as the decimal separator and all the others are group separators.

Removed the Localization part and checked if we can use `CurrencyInput`. We cannot.
Because that class can correct amounts if they are in the final state, but at CurrencyEntryBox they are typing. Managed to impement it, but the UX became horrible due to the random changes during typing.

soosr reviewed (commented):
Code LGTM. I will do a final testing later.

MarnixCroes reviewed (commented):
works ok.

one edge case (also on master, but maybe fixable in this refactor) for the `DualCurrencyEntryBox`: if the full entry is selected (double-click, or coming back preview tx dialog) and then click right arrow keystroke, the entry point is at the left, instead of the expected right side. Can only repro on the left currency entry (either BTC or USD) and when double-clicking not on the amount itself but next to it.

Didn't the up and down keystroke used to add one unit the amount? have 21 -> have the entry point there -> click up key -> 22.
It's not on master, but iirc we had that

AlexisKv commented:
If I have 0.123 btc, I cannot delete 0. It's just skipping to beginning.
So If I want to change 0 to 1, I need to write 10.123 and then delete zero, so it will looks like 1.123.

And I think that it should be impossible to delete `.` separator, when after separator are numbers.

soosr reviewed (changes requested):
- CTRL+A, CTRL+C, CTRL+V doesn't work
- I was able to paste text with Right click -> Paste (resulted in crash)
   - It seems like pasting doesn't care about any rule.
- When swapping input fields, can we select the whole input?
- Please add a tooltip to the swap button.
- Clicking on an amount suggestion just inserts the amount where the caret is instead of replacing the whole input field.

ichthus1604 commented:
@soosr 

> CTRL+A, CTRL+C, CTRL+V doesn't work

Fixed.

> I was able to paste text with Right click -> Paste (resulted in crash)
> It seems like pasting doesn't care about any rule.

Fixed.

> When swapping input fields, can we select the whole input?

Done.

> Please add a tooltip to the swap button.

Done.

> Clicking on an amount suggestion just inserts the amount where the caret is instead of replacing the whole input field.

Fixed.

MarnixCroes reviewed (changes requested):
- Enter doesn't work for continue when a currency amount is selected , it does on master
<details>
<summary>- error in log</summary>
<br>

```
[1] ERROR	CurrencyInputViewModel.PasteFromClipboardAsync (508)	System.ArgumentOutOfRangeException: startIndex cannot be larger than length of string. (Parameter 'startIndex')
   at System.String.ThrowSubstringArgumentOutOfRange(Int32 startIndex, Int32 length)
   at System.String.Substring(Int32 startIndex, Int32 length)
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.RemoveSelection() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 385
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.InsertRaw(String text) in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 151
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.PasteFromClipboardAsync() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 504
2024-01-31 09:32:33.236 [1] ERROR	CurrencyInputViewModel.PasteFromClipboardAsync (508)	System.ArgumentOutOfRangeException: startIndex cannot be larger than length of string. (Parameter 'startIndex')
   at System.String.ThrowSubstringArgumentOutOfRange(Int32 startIndex, Int32 length)
   at System.String.Substring(Int32 startIndex, Int32 length)
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.RemoveSelection() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 385
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.InsertRaw(String text) in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 151
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.PasteFromClipboardAsync() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 504
2024-01-31 09:32:33.588 [1] ERROR	CurrencyInputViewModel.PasteFromClipboardAsync (508)	System.ArgumentOutOfRangeException: startIndex cannot be larger than length of string. (Parameter 'startIndex')
   at System.String.ThrowSubstringArgumentOutOfRange(Int32 startIndex, Int32 length)
   at System.String.Substring(Int32 startIndex, Int32 length)
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.RemoveSelection() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 385
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.InsertRaw(String text) in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 151
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.PasteFromClipboardAsync() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 504
2024-01-31 09:32:33.773 [1] ERROR	CurrencyInputViewModel.PasteFromClipboardAsync (508)	System.ArgumentOutOfRangeException: startIndex cannot be larger than length of string. (Parameter 'startIndex')
   at System.String.ThrowSubstringArgumentOutOfRange(Int32 startIndex, Int32 length)
   at System.String.Substring(Int32 startIndex, Int32 length)
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.RemoveSelection() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 385
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.InsertRaw(String text) in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 151
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.PasteFromClipboardAsync() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 504
2024-01-31 09:32:38.714 [1] ERROR	CurrencyInputViewModel.PasteFromClipboardAsync (508)	System.ArgumentOutOfRangeException: startIndex cannot be larger than length of string. (Parameter 'startIndex')
   at System.String.ThrowSubstringArgumentOutOfRange(Int32 startIndex, Int32 length)
   at System.String.Substring(Int32 startIndex, Int32 length)
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.RemoveSelection() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 385
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.InsertRaw(String text) in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 151
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.PasteFromClipboardAsync() in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 504
```

</details>

- having an amount in clipboard, click on the amount entry then the amount tooltip to click on appears -> then minimize and then maximize wasabi -> the tooltip is at totally different place
- _Tab_ keystroke to jump to other entry doesn't work
- Both BTC and USD entries can be selected at the same time: select the right entry (ctrl a or part of it) click continue -> go back at preview tx -> left and right side are both selected
- click back at _Preview Transaction_, only part of the entry is selected

soosr reviewed (changes requested):
- Having the focus on the right side, then swap, the input on the left side will be selected, but the right side should be.

- After navigating back from the txn preview screen, a suggestion appeared on the top left of the dialog (couldn't repro). It seems like the suggestion flyout pops on the cursor's position, which is not correct, it should pop somewhere around the input field.

ichthus1604 commented:
@MarnixCroes @soosr 

> Enter doesn't work for continue when a currency amount is selected , it does on master

Fixed.

> Tab keystroke to jump to other entry doesn't work

Fixed.

>  error in log

Fixed.

ichthus1604 commented:
@wieslawsoltes 

> having an amount in clipboard, click on the amount entry then the amount tooltip to click on appears -> then minimize and then maximize wasabi -> the tooltip is at totally different place

> After navigating back from the txn preview screen, a suggestion appeared on the top left of the dialog (couldn't repro). It seems like the suggestion flyout pops on the cursor's position, which is not correct, it should pop somewhere around the input field.

This is caused by incorrect `PlacementMode` of the `FlyoutSuggestionBehavior`. However, there is this style:

```xaml
  <!-- Move Suggestion Behavior Flyout Placement -->
  <Style Selector="^[HorizontalContentAlignment=Left] /template/ behaviors|FlyoutSuggestionBehavior#SuggestionBehavior">
    <Setter Property="PlacementMode" Value="BottomEdgeAlignedLeft" />
  </Style>

  <Style Selector="^[HorizontalContentAlignment=Right] /template/ behaviors|FlyoutSuggestionBehavior#SuggestionBehavior">
    <Setter Property="PlacementMode" Value="BottomEdgeAlignedRight" />
  </Style>
```

In `CurrencyEntryBox.axaml`, which should have effect because the `CurrencyEntryBox.HorizontalContentAlignment` is effectively `Left` and `Right` respectively. Also, the behavior `x:Name` is correct as well.

Instead, it seems like the `PlacementMode` is the default, because it follows mouse pointer.

What could cause these to not have effect?

ichthus1604 commented:
@MarnixCroes 

> Both BTC and USD entries can be selected at the same time: select the right entry (ctrl a or part of it) click continue -> go back at preview tx -> left and right side are both selected

Fixed.

> click back at Preview Transaction, only part of the entry is selected

ichthus1604 commented:
@wieslawsoltes 

> Is FlyoutSuggestionBehavior styleable at all ?

base class `Behavior` is an `AvaloniaObject` so I assumed all behaviors are stylable? Is this correct?

Still, I fixed by code in `OnApplyTemplate()` and removed the XAML styles.

@MarnixCroes @soosr 

> having an amount in clipboard, click on the amount entry then the amount tooltip to click on appears -> then minimize and then maximize wasabi -> the tooltip is at totally different place

> After navigating back from the txn preview screen, a suggestion appeared on the top left of the dialog (couldn't repro). It seems like the suggestion flyout pops on the cursor's position, which is not correct, it should pop somewhere around the input field.

Fixed!

ichthus1604 commented:
known limitation: mouse dragging selection has been disabled. The way `TextBox` treates `SelectionStart` and `SelectionEnd` is incompatible with my implementation, so I removed it. We can re-enable it in a later PR if necessary.

soosr reviewed (changes requested):
- On send page, give the focus to the right side, press swap. The focus should have stayed on the right side, but it moved to the left side.

- The control isn't visible in settings:
![image](https://github.com/zkSNACKs/WalletWasabi/assets/16364053/51fefce0-5478-43dc-a0f1-61fc0a17eff0)

soosr commented:
> I wanted to test this and had a crash directly at entering the dialog. `UiClipboard.TryGetTextAsync (29) System.NullReferenceException` This is what I had in Clipboard: `https://github.com/zkSNACKs/WalletWasabi/pull/11966#pullrequestreview-1859481445`, but in fact any string creates a crash, maybe related to MacOs?
> 
> Full Logs:

Cannot repro on Win11

ichthus1604 commented:
@soosr 

> The control isn't visible in settings:

Fixed.

> Also broken for custom fee:

Fixed.

soosr commented:
Tested and works well enough for the release. Will review the code tomorrow.

> On send page, give the focus to the right side, press swap. The focus should have stayed on the right side, but it moved to the left side.

We discussed fixing it in the following PR as it is not important.

> I wanted to test this and had a crash directly at entering the dialog. `UiClipboard.TryGetTextAsync (29) System.NullReferenceException` This is what I had in Clipboard: `https://github.com/zkSNACKs/WalletWasabi/pull/11966#pullrequestreview-1859481445`, but in fact any string creates a crash, maybe related to MacOs?
> 
> EDIT: Even with an empty clipboard I have this bug
> 
> Full Logs:

@wieslawsoltes will debug the clipboard on osx.

yahiheb reviewed (commented):
The first time I run this branch I got a weird notification at the bottom right corner, saying that the changes of the settings will be applied after restarting Wasabi. I checked the settings and the changes message was there.

I don't know how to reproduce this, but it was weird.

soosr commented on WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:
```diff
@@ -0,0 +1,709 @@
+	public int? SelectionStart { get; set; }
+	public int? SelectionEnd { get; set; }
+
+	public bool HasSelection => SelectionEnd is { } && SelectionEnd is { };
```
One here should be `SelectionStart` right?

ichthus1604 replied:
Fixed

soosr commented on WalletWasabi.Fluent/Views/Wallets/CoinJoinSettingsView.axaml:
```diff
@@ -24,7 +24,7 @@
                     ToolTip.Tip="Coinjoin will not automatically start if the wallet balance is less than this.">
           <TextBlock Text="Auto-start coinjoin threshold" />
-          <c:CurrencyEntryBox Text="{Binding PlebStopThreshold}" CurrencyCode="BTC" />
+          <!--<c:CurrencyEntryBox Value="{Binding PlebStopThreshold}" />-->
```
This?

ichthus1604 replied:
Fixed.

soosr commented on WalletWasabi.Fluent/Controls/CurrencyEntryBox/CurrencyEntryBox.axaml.cs:
```diff
@@ -0,0 +1,375 @@
+					.Throttle(TimeSpan.FromMilliseconds(50), RxApp.MainThreadScheduler)
+					.Do(t =>
+					{
+						if (_isUpdating || t.First == t.Second || CurrencyFormat is null)
```
I don't get the `CurrencyFormat` here. It has a default value in this class and on ViewModel level it can never be null. Why are we checking for null here?

ichthus1604 replied:
Since `CurrencyFormat` is bindable, it might be bound to a null value in other cases. I think keeping this extra check here doesn't hurt.

soosr reviewed (commented):
Couldn't review this properly. Will do another round tomorrow.

- Exception when selection a part of Dust Threshold and copying:
```
2024-02-04 15:50:41.625 [1] ERROR       CurrencyInputViewModel.CopySelectionToClipboardAsync (504)      System.ArgumentOutOfRangeException: Index and length must refer to a location within the string. (Parameter 'length')
   at System.String.ThrowSubstringArgumentOutOfRange(Int32 startIndex, Int32 length)
   at System.String.Substring(Int32 startIndex, Int32 length)
   at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.CopySelectionToClipboardAsync() in WalletWasabi.Fluent\ViewModels\Wallets\Send\CurrencyConversion\CurrencyInputViewModel.cs:line 497

```

soosr commented:
I cannot reproduce those issues on Win11, @wieslawsoltes @ichthus1604 you guys have Mac, please test it deeply. @MarnixCroes Could you test this on Linux, just to make sure...

ichthus1604 commented on WalletWasabi.Fluent/Behaviors/FlyoutSuggestionBehavior.cs:
```diff
@@ -12,6 +13,8 @@ namespace WalletWasabi.Fluent.Behaviors;
 
 public class FlyoutSuggestionBehavior : AttachedToVisualTreeBehavior<Control>
 {
+	public static readonly StyledProperty<string> NameProperty = AvaloniaProperty.Register<FlyoutSuggestionBehavior, string>(nameof(Name));
```
Removed.

ichthus1604 commented on WalletWasabi.Fluent/Models/UI/UiClipboard.cs:
```diff
@@ -9,6 +10,20 @@ public partial class UiClipboard
 {
 	public async Task<string> GetTextAsync() => await ApplicationHelper.GetTextAsync();
 
+	public async Task<string?> TryGetTextAsync()
+	{
+		try
+		{
+			return await GetTextAsync();
+		}
+		catch (Exception ex)
+		{
+			Logger.LogError(ex);
+		}
+
+		return null;
+	}
```
No. This is used because I've noticed there are times when `GetTextAsync()` crashes inside an Avalonia method.
I decided to just ignore the exception and retry later.

soosr reviewed (changes requested):
It doesn't work on osx:

- Entering any number for the first time into the field once, ends up being entered twice.
   - Moving the caret with left and right arrow is glitchy at this state.
   - Entering one additional number results in exception:
   ```
   2024-02-06 09:27:13.698 [1] ERROR	CurrencyInputViewModel.Insert (84)	System.ArgumentOutOfRangeException: Index and length must refer to a location within the string. (Parameter 'length')
  at System.String.ThrowSubstringArgumentOutOfRange(Int32 startIndex, Int32 length)
  at System.String.Substring(Int32 startIndex, Int32 length)
  at WalletWasabi.Fluent.ViewModels.Wallets.Send.CurrencyConversion.CurrencyInputViewModel.Insert(String text) in WalletWasabi.Fluent/ViewModels/Wallets/Send/CurrencyConversion/CurrencyInputViewModel.cs:line 62
   ```
   - Removing one number cause a same exception.
- Pressing `.` results in `0..` which probably the same issue as before.
- (osx, windows) Caret's position should be the very end, when clicking on the amount text on the top right of Send dialog.
- Clicking on the suggested amount flyout doesn't work for BTC field. (work well for USD)
- Entering a amount into USD field, then clicking on the BTC field (the input is not selected) then pressing BACKSPACE removes the whole input.
   - Moving the caret left then right solves the issue and backspace work as expected.

soosr commented:
- **Mac** Clicking on the suggested amount Flyout doesn't work on the BTC field (works well on USD)
- **Mac** CMD + V doesn't work. CMD + V + V works for pasting
- **Mac** CMD + A doesn't work. CMD + A + A works for selecting all
   - **Mac** Pasting `0.00023` then selecting all, doesn't select the whole input, last char is missing.
- **Mac** Selecting with mouse doesn't work. (double click, select all works)
- **Mac** Cannot start input with `0`.
   - Actually cannot manually type in `0` at all.
- **Windows** After Entering `0` the caret is visually in the wrong position (its in the beginning), although adding new numbers goes after 0.
   - Actually if you start typing, you will see the caret most of the time is before the last char.
   - Can repro on the USD field too.
- **Mac** Selecting with shift + arrow, the first action doesn't do anything.

SuperJMN commented:
I've tested this in Linux (444b6295) and by far, the worst UX problem is this:
- Given blank input, type any digit. Delete it. Type any digit again. The caret will position incorrectly on the left side of the just typed digit.
![image](https://github.com/zkSNACKs/WalletWasabi/assets/3109851/ca896817-b73b-484f-a6e3-3aab373638aa)

soosr commented:
Converting to draft due to mac issues
